### PR TITLE
Mp3parser

### PR DESCRIPTION
**Pull request by svitvitskiy**
Needed to be able to mux raw mp3 files into container. See feature request #246

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -31,18 +36,15 @@ JCodec is free software distributed under FreeBSD License.
 
 * Video
     * Improve H.264 encoder: CABAC, rate control;
-    * Performance optimize H.264 decoder;
-    * Implement H.264 encoder on RenderScript;
```
Comment (zhuker) [inline README.md]: RenderScript was a good idea

File: README.md [hunk 2]
```diff
@@ -65,33 +78,48 @@ For JDK
 </dependency>
 ```
 
-For Android
-```xml
-android {
-        configurations.all {
-        resolutionStrategy.force 'com.google.code.findbugs:jsr305:3.0.2'
```
Comment (zhuker) [inline README.md]: this part forces findbugs to be of the version we use

File: README.md [hunk 3]
```diff
@@ -65,33 +78,48 @@ For JDK
...
 
+# Performance / quality considerations
+Because JCodec is a pure Java implementation please adjust your performance expectations accordingly. We usually make the best effort to write efficient code but despite this the decoding will typically be an order of magnitude slower than the native implementations (such as FFMpeg). We are currently looking into implementing performance-critical parts in OpenCL (or RenderScript for Android) but the ETA is unknown.
+
+Expect the encoded quality/bitrate for h.264 (AVC) to be so much worse compared to the well known native encoders (such as x264). This is because very little work has been put so far into developing the encoder and also because encoders usually trade speed for quality, speed is something we don't have in Java, hence the quality. Again we may potentially fix that in the future by introducing OpenCL (RenderScript) code but at this point it's an unknown.
+
+That said the decode quality should at the industry level. This is because the decoding process is usually specified by the standard and the correct decoder implementations are expected to produce bit-exact outputs.
```
Comment (zhuker) [inline README.md]: should _be_ at the industry level.

File: src/main/java/org/jcodec/api/transcode/Transcoder.java [hunk 4]
```diff
@@ -201,6 +199,14 @@ public void finalFlushQueues() throws IOException {
                         }
                     }
                 }
+            } else {
+                for (AudioFrameWithPacket audioFrame : audioQueue) {
```
Comment (zhuker) [inline src/main/java/org/jcodec/api/transcode/Transcoder.java]: `AudioFrameWithPacket` - maybe lets call it `AudioSampleBuffer` like Apple Core Media does?
https://developer.apple.com/documentation/coremedia/cmsamplebuffer-u71
Basic idea is:
SampleBuffer == data + timing + extra metadata

File: src/main/java/org/jcodec/containers/mp3/MPEGAudioDemuxer.java [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+                readMoreData();
+            if (!readBuffer.hasRemaining()) {
+                eof = true;
+                break;
+            }
+            runningFour <<= 8;
+            runningFour |= readBuffer.get() & 0xff;
+            ++total;
+        }
+        Logger.warn(String.format("[mp3demuxer] Skipped %d byts of junk", total));
```
Comment (zhuker) [inline src/main/java/org/jcodec/containers/mp3/MPEGAudioDemuxer.java]: > byts

bytes

File: README.md (hunk 1 above)
Comment (svitvitskiy) [reply README.md]: Agree, but realistically nobody is going to do it in the next six months. I am don't want to set the expectations high.

File: src/main/java/org/jcodec/api/transcode/Transcoder.java (hunk 4 above)
Comment (svitvitskiy) [reply src/main/java/org/jcodec/api/transcode/Transcoder.java]: there is a class org.jcodec.common.model.AudioBuffer that contains data + associated audio format information. This Class is the internal of the transcoder to carry 2 pieces, i.e. the decoded AudioBuffer + the packet. I would prefer to leave the name as is, or replace it with Pair (_2) AudioBuffer, Packet. Let me know.

**Pull request merged**